### DISCUSSION TITLE
Proposal: Override return in function calls

### DISCUSSION THREAD

**Discussion by TheUnlocked** [dotnet/csharplang#1342] [Language Ideas]
These might be existing extension functions to make code cleaner:
```cs
public static Foo BarExt(this Foo obj){
     obj.Bar();
     return obj;
}
public static Foo BazExt(this Foo obj){
     obj.Baz();
     return obj;
}
public static Foo WompExt(this Foo obj){
     obj.Womp();
     return obj;
}
```
Which allows one to write:
```cs
Pow(obj.BarExt().BazExt().WompExt());
```
Instead of
```cs
obj.Bar();
obj.Baz();
obj.Womp();
Pow(obj);
```

Making all of those extensions is really annoying though, and it can also be somewhat misleading. Typically, a function that returns a modified version of the `this` object will return a new instance of the object, whereas those extension functions just modify it. It's not necessarily intuitive and it may require a lot of repetitive code, either in the extensions or in the call. There should be a better way.

So here is my proposed method of making one-line code in an understandable way:
```cs
Pow(obj.Bar(return obj).Baz(return obj).Womp(return obj));
```

By using the return keyword at the end of a function call, it would ignore the return value of the function (including void), and that call would instead return whatever object was put in that call.

An alternative to `return` could be `returns`, though that isn't an existing keyword so I don't know if that would cause any issues. The `returns` keyword _shouldn't_ cause issues with back-compatibility, though maybe I'm missing something there.

**Comment by CyrusNajmabadi** [comment]
Here is mine:

```c#
obj.Bar(); obj.Baz(); obj.Womp(); Pow(obj); // vs
Pow(obj.Bar(return obj).Baz(return obj).Womp(return obj));
```

> So here is my proposed method of making one-line code in an understandable way:

There's already a way to make things one line.  And it has the virtue of being even shorter, even simpler, even more understandable, and can be done today without any language changes :)

**Comment by TheUnlocked** [comment]
It is one _line_, but it's not one _statement_, and that can matter.
Imagine this scenario:
```cs
if (obj.Bar(return obj).Equals(obj2))
     obj2 = obj.Baz(return obj).Womp(return obj).Quack(obj2);
// vs
obj.Bar();
if (obj.Equals(obj2)) {
     obj.Baz(); obj.Womp();
     obj2 = obj.Quack(obj2);
}
```

**Comment by CyrusNajmabadi** [comment]
I don't see anything wrong with teh second form :)

If you have long and complex functionality you want to encapsulate as an expression, then extracting an extension, or making a local function, seems like it would do nicely to solve the problem.

**Comment by HaloFour** [comment]
Personally I don't understand the desire to turn everything into a single-line expression.

Putting that aside, I also don't follow the suggested syntax.  Where exactly is this "return override" supposed to appear within the parameter list when the method actually accepts parameters, especially `params` parameters?  It also appears to be an expression form of the `return` statement, which clashes syntactically with [return expressions](https://github.com/dotnet/csharplang/issues/867) which achieve something else entirely.

**Comment by TheUnlocked** [comment]
If `return` is already planning on being used, the alternative `returns` could be used instead.

As for where it appears, it wouldn't actually be in the parameter list. Here's an example:

```cs
public int Foo(string s, params int[] nums){
     int val = 0;
     // Something
     return val;
}

int q = Foo("Test", x, 5, 2, 12, 555, returns x);
```

**Comment by kameko** [comment]
Isn't this basically ``in``/``out``?

I agree with HaloFour, I don't know what the deal is with this recent trend of this issues section being bombarded by people who think this is Perl, but this isn't Perl, it's C#, a language with a philosophy of writing code that's purposefully mildly-verbose for the sake of readability and clarity.
Also, recommending new keywords is the last thing you should do. The team hates new keywords and tries to re-purpose them where it seems reasonable.

**Comment by DavidArno** [comment]
I'm completely confused as to how this proposal works for your last example:

> `int q = Foo("Test", x, 5, 2, 12, 555, returns x);`

Is `q` assigned the value of `x`, or the return value from `Foo`?

**Comment by TheUnlocked** [comment]
Whatever `x` is after calling the function.

**Comment by DavidArno** [comment]
@TheUnlocked,

And what does `x` contain after calling the function? Let's try another example:
```cs
int One() => 1;

var x = 2;
var y = One(return x);
```
After the last line is called, what would the value of `x` and `y` be?

**Comment by TheUnlocked** [comment]
That would set both x and y to 2.

I'm going to concede this proposal as others' arguments are compelling, but I'll leave it open for further discussion.

**Comment by Richiban** [comment]
Isn't it a lot better to have an actual method chaining syntax? Although, now that I think about it, I think that's been turned down in the past. Examples might have been:

```csharp

obj.Foo()
   ..Bar()
   ..Baz();

```
